### PR TITLE
Start using Project Metadata to track non-core Project fields.

### PR DESCRIPTION
**Pull request by devinleighsmith**

### CODE REVIEW

File: backend/api/Areas/Project/Mapping/Dispose/ProjectMap.cs [hunk 1]
```diff
@@ -1,4 +1,5 @@
 using Mapster;
```
Comment (devinleighsmith) [inline backend/api/Areas/Project/Mapping/Dispose/ProjectMap.cs:1]: Overall the idea is to push the decision whether to map the metadata or not to the mapper, that way a controller can still call the project service without having to deal with the performance cost of serialization.

File: backend/api/Areas/Project/Mapping/Dispose/ProjectMap.cs [hunk 2]
```diff
@@ -57,6 +58,7 @@ public void Register(TypeAdapterConfig config)
                 .Map(dest => dest.OfferAmount, src => src.OfferAmount)
                 .Map(dest => dest.Tasks, src => src.Tasks)
                 .Map(dest => dest.ProjectAgencyResponses, src => src.Responses)
+                .BeforeMapping((src, dest) => JsonConvert.PopulateObject(src.Metadata ?? "{}", src)) //use the metadata object to populate fields not stored in DB.
```
Comment (devinleighsmith) [inline backend/api/Areas/Project/Mapping/Dispose/ProjectMap.cs:61]: This allows the mapping to behave the same way it did before these fields were removed, all fields will be mapped directly from the entity to the model since the entity will be populated with the metadata before the mapping occurs.

Note: populateObject respects the field annotations in the Project object, so only [JsonProperty] fields will be populated.

File: backend/api/Areas/Project/Mapping/Dispose/ProjectMap.cs [hunk 3]
```diff
@@ -100,6 +102,7 @@ public void Register(TypeAdapterConfig config)
                 .Map(dest => dest.OfferAmount, src => src.OfferAmount)
                 .Map(dest => dest.Tasks, src => src.Tasks)
                 .Map(dest => dest.Responses, src => src.ProjectAgencyResponses)
+                .AfterMapping((src, dest) => dest.Metadata = JsonConvert.SerializeObject(dest)) //Map all non-ignored fields into the metadata.
```
Comment (devinleighsmith) [inline backend/api/Areas/Project/Mapping/Dispose/ProjectMap.cs:105]: same idea, map the fields as per usual, allowing entity logic requiring fields that are stored in metadata to perform required checks. After mapping, serialize all [JsonProperty] fields to Metadata.

File: backend/api/Areas/Project/Mapping/Search/ProjectMap.cs [hunk 4]
```diff
@@ -39,6 +40,7 @@ public void Register(TypeAdapterConfig config)
                 .Map(dest => dest.CreatedOn, src => src.CreatedOn)
                 .Map(dest => dest.CreatedBy,
                     src => src.CreatedById != null ? src.CreatedBy.DisplayName : null)
+                .BeforeMapping((src, dest) => JsonConvert.PopulateObject(src.Metadata ?? "{}", src))
```
Comment (Fosol) [inline backend/api/Areas/Project/Mapping/Search/ProjectMap.cs:43]: The only concern I have here for the search is that we're returning an array of projects.  Ideally we won't do any extra serialization here if we don't need to.

File: backend/dal/Services/Concrete/ProjectService.cs [hunk 5]
```diff
@@ -780,6 +781,7 @@ public async System.Threading.Tasks.Task<Project> SetStatusAsync(Project project
             // Update a project
             originalProject.StatusId = toStatus.Id;
             originalProject.Status = toStatus;
+            originalProject.Metadata = project.Metadata;
```
Comment (devinleighsmith) [inline backend/dal/Services/Concrete/ProjectService.cs:784]: Something to think about: From what I could tell, it shouldn't be necessary to populate originalProject with metadata values either in the update or setStatus functions as those values don't seem to be required for business logic. Correct me if I'm wrong.

File: backend/entities/Project.cs [hunk 6]
```diff
@@ -1,11 +1,15 @@
...
+using System.ComponentModel.DataAnnotations.Schema;
+using System.Text.Json.Serialization;
+using Newtonsoft.Json;
 
 namespace Pims.Dal.Entities
 {
     /// <summary>
     /// Project class, provides an entity for the datamodel to manage projects.
     /// </summary>
+    [JsonObject(MemberSerialization.OptIn)]
```
Comment (devinleighsmith) [inline backend/entities/Project.cs:12]: Could also have marked all of the DB properties as JsonIgnore, but I think that it makes more sense to keep the property attributes grouped like this.

File: backend/entities/Pims.Dal.Entities.csproj [hunk 7]
```diff
@@ -8,6 +8,7 @@
   </PropertyGroup>
   <ItemGroup>
     <PackageReference Include="Microsoft.EntityFrameworkCore" Version="3.1.5" />
+    <PackageReference Include="Newtonsoft.Json" Version="12.0.3" />
```
Comment (Fosol) [inline backend/entities/Pims.Dal.Entities.csproj:11]: We shouldn't need Newtonsoft, as we don't use it anywhere else for serialization.  Use the built in System.Text.Json.

File: backend/entities/Project.cs [hunk 8]
```diff
@@ -118,6 +122,8 @@ public class Project : BaseEntity
         /// <summary>
         /// get/set - When the project was externally marketed.
         /// </summary>
+        [NotMapped]
+        [JsonProperty]
```
Comment (Fosol) [inline backend/entities/Project.cs:126]: I'd extract all these extra properties and create a `DisposalProject` in the `/Models` namespace within this library.

File: backend/tests/core/Entities/ProjectHelper.cs [hunk 9]
```diff
@@ -62,6 +62,7 @@ public static Entity.Project CreateProject(int id, Entity.Agency agency = null,
                 PublicNote = $"publicNote-{id}",
                 PrivateNote = $"privateNote-{id}",
                 UpdatedOn = DateTime.UtcNow,
+                Metadata = "{offerAmount: 123}",
```
Comment (Fosol) [inline backend/tests/core/Entities/ProjectHelper.cs:65]: Technically that isn't valid JSON, but I don't know if it'll cause any issue. `{"offerAmount": 123}`

File: backend/dal/Services/Concrete/ProjectService.cs (hunk 5 above)
Comment (Fosol) [reply backend/dal/Services/Concrete/ProjectService.cs:784]: I'd think you'd need it in `Add` and `Update`, otherwise you'll never have values there.  Since `SetStatus` also updates, you'll probably need it there too.

There are some possible business rules related to required date values before proceeding between states, but we may only enforce them currently on the frontend.  Ideally we'd enforce in the backed too.

File: backend/api/Areas/Project/Mapping/Dispose/ProjectMap.cs (hunk 1 above)
Comment (Fosol) [reply backend/api/Areas/Project/Mapping/Dispose/ProjectMap.cs:1]: Perfect!

File: backend/tests/core/Entities/ProjectHelper.cs (hunk 9 above)
Comment (devinleighsmith) [reply backend/tests/core/Entities/ProjectHelper.cs:65]: The test passes as-is.

File: backend/entities/Project.cs (hunk 8 above)
Comment (devinleighsmith) [reply backend/entities/Project.cs:126]: If we go that route I think I'd need to update the Project Service to operate on the new DisposalProject instead of the Project object to keep the mappings working the way they do now. Alternatively I could map the model to the Project, and then convert the project to a DisposalProject by deserializing the metadata right after it was serialized. I'm not a fan of either of those options, but maybe I don't understand the suggestion.

File: backend/entities/Pims.Dal.Entities.csproj (hunk 7 above)
Comment (devinleighsmith) [reply backend/entities/Pims.Dal.Entities.csproj:11]: [JsonObject(MemberSerialization.OptIn)] is only part of Newtonsoft, let me see if there is an equivalent in System.Text.Json.Serialization

File: backend/api/Areas/Project/Mapping/Search/ProjectMap.cs (hunk 4 above)
Comment (devinleighsmith) [reply backend/api/Areas/Project/Mapping/Search/ProjectMap.cs:43]: I'll test tasking this out and see if it breaks anything.

File: backend/entities/Pims.Dal.Entities.csproj (hunk 7 above)
Comment (Fosol) [reply backend/entities/Pims.Dal.Entities.csproj:11]: I doubt our serialization would obey a Newtonsoft specific attribute.

File: backend/entities/Project.cs (hunk 8 above)
Comment (Fosol) [reply backend/entities/Project.cs:126]: You've essentially lost some of the reason to use the `metadata` property by keeping `NotMapped` properties on the entity `Project`.  It makes more sense to have a specific model for each given project type.  Thus creating a `DisposalProject` model that has these properties which can be used to work with the `Metadata` properties.

Currently yes it would require updating the `ProjectService` functions, but could theoretically be handled via generics.

Comment (devinleighsmith) [reply backend/entities/Project.cs:126]: maybe a partial would work here?

Comment (Fosol) [reply backend/entities/Project.cs:126]: A temporary option, but in truth isn't any different.

Comment (devinleighsmith) [reply backend/entities/Project.cs]: I think I'll hold off on this for now since this at least alleviates some of the pain in the database. If I have time I'll see if I can enhance this further.

File: frontend/src/features/projects/dispose/ProjectDisposeLayout.tsx [hunk 10]
```diff
@@ -134,7 +134,6 @@ const ProjectDisposeLayout = ({ match, location }: { match: Match; location: Loc
           {getProjectRequest?.isFetching !== true ? (
             <Container fluid className="step-content">
               <Switch>
-                )}
```
Comment (asanchezr) [inline frontend/src/features/projects/dispose/ProjectDisposeLayout.tsx:137]: Is this a bug fix for a separate issue? This frontend file doesn't seem related to the PR work...
